### PR TITLE
[deckhouse-controller] Fix corner cases in d8-cluster-configuration webhook

### PR DESCRIPTION
**Pull request by armanokka**
## Description
Fix corner cases in d8-cluster-configuration webhook.

## Why do we need it, and what problem does it solve?
fix corner cases in d8-cluster-configuration webhook

## Why do we need it in the patch release (if we do)?
no need

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: fix
summary: Fixed corner cases in d8-cluster-configuration webhook.
impact_level: default
```

### CODE REVIEW

File: deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go [hunk 1]
```diff
@@ -148,61 +149,168 @@ func getKubernetesEndpointsCount(ctx context.Context, cli client.Client) (int, e
...
 	}
 
-	verNew, err := parseVersion(newVersion)
-	if err != nil {
-		return nil, fmt.Errorf("failed to parse new version: %w", err)
+	selectedChecker = minorSubCheck
+
+	// Resolve oldVersion: if it's "Automatic", get actual version from secret
+	var oldVersionSemver *semver.Version
+	var oldVersionStr string
```
Comment (090809) [inline deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go]: Why we need this string?

File: deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go [hunk 2]
```diff
@@ -148,61 +149,168 @@ func getKubernetesEndpointsCount(ctx context.Context, cli client.Client) (int, e
...
+		// we cannot determine the actual version that was used, so we allow the change.
+		// This can happen during initial cluster setup or if secret is incomplete.
+		if !exists {
+			return allowResult(nil)
+		}
+
+		maxUsedVersion := strings.TrimSpace(string(maxUsedVersionB64))
+		oldVersionStr = maxUsedVersion
+		var err error
+		oldVersionSemver, err = parseVersion(maxUsedVersion)
```
Comment (090809) [inline deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go]: Already trimmed in parseVersion

File: deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go [hunk 3]
```diff
@@ -148,61 +149,168 @@ func getKubernetesEndpointsCount(ctx context.Context, cli client.Client) (int, e
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse maxUsedControlPlaneKubernetesVersion: %w", err)
+		}
+	} else {
+		oldVersionStr = oldVersion
```
Comment (090809) [inline deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go]: why we need this?

File: deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go [hunk 4]
```diff
@@ -148,61 +149,168 @@ func getKubernetesEndpointsCount(ctx context.Context, cli client.Client) (int, e
...
-	if exists {
-		maxUsedVersion := string(maxUsedVersionB64)
-		verMax, err := parseVersion(maxUsedVersion)
-		if err == nil {
-			if verMax.Major() > verNew.Major() || (verMax.Major() == verNew.Major() && verMax.Minor() > verNew.Minor()+1) {
-				return rejectResult(fmt.Sprintf("can not downgrade kubernetes version more than 1 minor version. maxUsedControlPlaneKubernetesVersion=%s newKubernetesVersion=%s", maxUsedVersion, newVersion))
+	// Additional check: maxUsedControlPlaneKubernetesVersion
+	// This prevents downgrade below the maximum version that was ever used in the cluster.
+	// Only check this for non-Automatic newVersion, as Automatic already uses deckhouseDefaultKubernetesVersion.
+	if newVersion != "Automatic" {
```
Comment (090809) [inline deckhouse-controller/pkg/apis/deckhouse.io/validation/validate_cluster_configuration.go]: This scenario already catched via selectedChecker (it may be minorSubCheck (as default)), no chance for getting this

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @armanokka. I've started the workflow for label 'e2e/run/yandex-cloud'...

  :fast_forward: `e2e: Yandex.Cloud, Containerd, Kubernetes 1.32` for `deckhouse:rewrite-webhook-d8-cluster-configuration` [started](https://github.com/deckhouse/deckhouse/actions/runs/20591251617).

**Comment by deckhouse-BOaTswain** [comment]
:red_circle: `e2e: DVP` for `deckhouse:rewrite-webhook-d8-cluster-configuration` [failed](https://github.com/deckhouse/deckhouse/actions/runs/20910898037) in 2m19s.




E2e for dvp WithoutNAT;containerd;1.32 was failed. Use:
  

  `/e2e/abort dvp;WithoutNAT;containerd;1.32 20910898037 20910898037-1-con-1-32 /sys/deckhouse-oss/install:pr17342 ` - for abort failed cluster




<details><summary>Workflow details (1 job failed)</summary>


:red_circle: `e2e: DVP, Containerd, Kubernetes 1.32` failed in 1m10s.

</details>

**Comment by deckhouse-BOaTswain** [comment]
:green_circle: `e2e: Yandex.Cloud` for `deckhouse:rewrite-webhook-d8-cluster-configuration` [succeeded](https://github.com/deckhouse/deckhouse/actions/runs/20915794424) in 36m18s.
<details><summary>Workflow details</summary>



Yandex.Cloud-WithoutNAT-Containerd-1.32 - Connection string: `ssh  redos@158.160.34.91`


:green_circle: `e2e: Yandex.Cloud, Containerd, Kubernetes 1.32` succeeded in 35m3s.

</details>
